tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/BsonSerialization/BsonSerializationHelper.cs

```diff
@@ -0,0 +1,232 @@
+            var proxyResult = DuckType.GetOrCreateProxyType(typeof(IBsonSerializerLookupProxy), bsonSerializerType);
+            if (!proxyResult.Success)
+            {
+                return null;
+            }
+
+            var bsonSerializerLookupProxy = (IBsonSerializerLookupProxy)proxyResult.CreateInstance(null!);
+
+            proxyResult = DuckType.GetOrCreateProxyType(typeof(IBsonSerializationContextProxy), bsonSerializationContextType);
+            if (!proxyResult.Success)
+            {
+                return null;
+            }
+
+            var bsonSerializationContextProxy = (IBsonSerializationContextProxy)proxyResult.CreateInstance(null!);
+
+            proxyResult = DuckType.GetOrCreateProxyType(typeof(JsonWriterSettingsProxy), jsonWriterSettingsType);
+            if (!proxyResult.Success)
+            {
+                return null;
+            }
+
+            var jsonWriterSettingsProxy = (JsonWriterSettingsProxy)proxyResult.CreateInstance(null!);
+
+            // Create JSonWriter
+            var jsonWriterCtor = jsonWriterType.GetConstructor(new[] { typeof(TextWriter), jsonWriterSettingsType })!;
+
+            DynamicMethod createJsonWriterMethod = new DynamicMethod(
+                $"MongoJsonWriterSerializer",
+                jsonWriterType,
+                parameterTypes: new[] { typeof(TextWriter), typeof(object) },
+                typeof(DuckType).Module,
+                true);
+
+            ILGenerator createJsonWriterIl = createJsonWriterMethod.GetILGenerator();
+            createJsonWriterIl.Emit(OpCodes.Ldarg_0);
+            createJsonWriterIl.Emit(OpCodes.Ldarg_1);
+            // createJsonWriterIl.Emit(OpCodes.Castclass, jsonWriterSettingsType); // Not technically necessary
+            createJsonWriterIl.Emit(OpCodes.Newobj, jsonWriterCtor);
+            createJsonWriterIl.Emit(OpCodes.Ret);
+
+            var createJsonWriterFunc = (Func<TextWriter, object, object>)createJsonWriterMethod.CreateDelegate(typeof(Func<TextWriter, object, object>));
+
+            var bsonSerializationArgsCtor = bsonSerializationArgsType.GetConstructor(new[] { typeof(Type), typeof(bool), typeof(bool) })!;
+
+            // Create BsonSerializationArgs
+            DynamicMethod createBsonSerializationArgs = new DynamicMethod(
+                $"MongoBsonSerializationArgsType",
+                typeof(object), // boxed bsonSerializationArgsType
+                parameterTypes: new[] { typeof(Type) },
+                typeof(DuckType).Module,
+                true);
+
+            ILGenerator bsonWriterIl = createBsonSerializationArgs.GetILGenerator();
+            bsonWriterIl.Emit(OpCodes.Ldarg_0);
+            bsonWriterIl.Emit(OpCodes.Ldc_I4_0);
+            bsonWriterIl.Emit(OpCodes.Ldc_I4_0);
+            bsonWriterIl.Emit(OpCodes.Newobj, bsonSerializationArgsCtor);
+            bsonWriterIl.Emit(OpCodes.Box, bsonSerializationArgsType);
+            bsonWriterIl.Emit(OpCodes.Ret);
```

Comment:
Yikes, I wish we had a ducktype helper for those

Reply:
Yeah 🙁  We have a helper for when we want to call a no-args constructor, but not for these. Not sure we can do anything with DuckTyping directly for constructors 🤔